DR-2826 Bulk file ingest mode

nmalfroy opened the pull request:
OK, this is kind of a big one but here we go.  This PR introduces a bulk mode for GCP specifically for dealing with large numbers of files

The problem stems from the fact that the current ingest mechanism splits all file ingests into individual subflights. There's a definite advantage to this when there are smaller numbers of files (e.g. < 10K) since it has all sort so safeguards that are easier to read (by virtue of living in individual steps).  The draw back however is that it does not scale.  We see timings getting exponentially worse as we add more files.

All of this is to say: we are not getting rid of the current ingest mechanism but are adding an new mode of ingest which is specified by setting `bulkMode: true` in the various file and data ingest APIs. (the default is false)

I took pains in the PR to reduce chattiness with the postgres db (which is where a ton of bottleneck comes from) and to reduce where I could the amount of data needed to be stored in the stairway db (by explicitly unsetting working map values so that they don't keep getting copied on subsequent steps)

The main crux of the PR is in the `IngestBulkGcpStep` step.  This step contains most of the logic needed to ingest files:
- it copies files (added some logic to make sure that files aren't copied if they already exist in the target)
- it bulk ingest directory information into firestore (logic and time savings there)
- it bulk ingests file metadata into firestore (logic and time savings here)
- it foregoes the load / load lock tables in favor of storing a blob of file load info in the stairway DB and just using that to populate the load history table

As you can imagine, that's a whole lot moving parts which is why this PR is so big.  I tried really hard to minimize the amount of new integration tests added and tried to focus my efforts on unit and connected tests.

From a metrics standpoint, it's worth noting: I'm able to do an ingest of 100K files in self hosted mode in ~1h10minutes (down from roughly 20 hours)

Drivebys:
- I changed the query logging for BQ queries to debug mode since there were some truly massive log statements and it didn't seem useful

nmalfroy committed:
Method to get a list of all parent dirs given a path

nmalfroy committed:
Add bulk operations for inserting data into Firestore

nmalfroy committed:
Make GCP file copy skip if target file already exists

nmalfroy committed:
Make sure file id calculation works with null md5

nmalfroy committed:
Add ability to create LoadHistoryIterator from list

(It's used to populate the load history table and we'll be storing that info in the flight map in bulk mode)

nmalfroy committed:
Create a new base step which runs a full bulk ingest

I attempted to mimic what the flights do in order to minimize the amount of fragmentation from running scattered ingests and to reduce the chattiness with the DB in favor of large object storage in the stairway DB

nmalfroy committed:
Make it so that the load history table can be loaded from bulk mode

nmalfroy committed:
Make things less ingest-type specific

nmalfroy committed:
Wire in bulk mode to ingest flights

nmalfroy committed:
Add bulk mode options to API

nmalfroy committed:
Add support for bulk loading when ingesting in array mode

nmalfroy committed:
Add integration testing for the various ingest modes

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpStep.java:
```diff
@@ -0,0 +1,520 @@
+      List<CopyResult> copyResults =
+          doFileCopy(workingMap, loadModels, fileIdsByPath, bucketResource);
+
+      fsFileInfos.addAll(
+          copyResults.stream()
+              .filter(r -> Objects.isNull(r.error()) && Objects.nonNull(r.fsFileInfo()))
+              .map(CopyResult::fsFileInfo)
+              .toList());
```
Instead of `.addAll(stream().....toList())` I think it's better to write `stream()...forEach(fsFileInfos:add)` because that avoids creating a temporary result of the entire array list in memory.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpStep.java:
```diff
@@ -0,0 +1,520 @@
+
+      fsFileInfos.addAll(
+          copyResults.stream()
+              .filter(r -> Objects.isNull(r.error()) && Objects.nonNull(r.fsFileInfo()))
```
The `isNull` and `nonNull` methods are intended to be used as functional arguments If you want to compare against `null` directly, `!= null` is better IMO.
```suggestion
              .filter(r -> r.error() == null && r.fsFileInfo() != null)
```

nmalfroy committed:
Spotless and PR comments

nmalfroy committed:
Make the query log in debug mode

These debug messages can get very big so best not to overdo it unless we actually need to debug a query

nmalfroy commented on src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpStep.java:
```diff
@@ -0,0 +1,520 @@
+
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    // Do not to undo
```
typo

nmalfroy commented on src/main/java/bio/terra/service/filedata/FSId.java:
```diff
@@ -0,0 +1,36 @@
+package bio.terra.service.filedata;
```
Note: can revert this commit

okotsopoulos commented on src/test/java/bio/terra/service/filedata/FileMetadataUtilsTest.java:
```diff
@@ -149,6 +149,13 @@ public void testUniqueNewPaths() {
   }
 
+  @Test
+  public void extractDirectoryPathsTest() {
```
Nice to see a concrete example of what this utility does, would you mind expanding to include base case behavior if appropriate (i.e. empty string passed in)?

nmalfroy replied:
oh right on

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -81,13 +81,59 @@ public void createDirectoryEntry(Dataset dataset, FireStoreDirectoryEntry newEnt
+   * @param dataset The dataset being ingested into
+   * @param loadTag the load tag of the current ingest
+   * @param directories All directories to upsert
+   * @return A list of booleans the same size as the directories parameter where if an entry is
```
Looks like this is a void method, should this @return annotation be removed?  I'm not 100% clear on what the expected end state of this method is.

nmalfroy replied:
derp good call. that was copy pasta. I'll make it return the same conflict map for symmetry (though we don't really need it since this is just for directories)

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -134,6 +138,112 @@ public void createDirectoryEntry(
+   * @param collectionId the id of the collection being updated
+   * @param loadTag the load tag of the current ingest
+   * @param directories All directories to upsert
+   * @return A list of booleans the same size as the directories parameter where if an entry is
```
I'm going through this PR commit by commit, so this may become clearer when I see how this is ultimately used :)

Had you considered returning instead a `Map<String, Boolean>` which maps directories to their upsert status?  That may clarify the values and guard against a future situation where this information is passed around and separated from the directory list.

nmalfroy replied:
I ended up getting rid of this method an consolidating with the one below (the boolean was only used for tests)

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -134,6 +138,112 @@ public void createDirectoryEntry(
+                      if (documentSnapshot.exists()) {
+                        FireStoreDirectoryEntry existingEntry =
+                            documentSnapshot.toObject(FireStoreDirectoryEntry.class);
+                        // If the file exists with the same load tag, we assume that it's from
```
Appreciate this comment :)

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -134,6 +138,112 @@ public void createDirectoryEntry(
+        .filter(Objects::nonNull)
+        .collect(Collectors.toMap(IdConflict::attemptedId, IdConflict::existingId));
+  }
+  // Need to declare the record class out here and not inline in the method sine that makes spotbugs
```
Lol! I think I've encountered this with Spotbugs as well.

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreFileDao.java:
```diff
@@ -169,7 +192,7 @@ void deleteFilesFromDataset(
         firestore,
         collectionId,
-        DELETE_BATCH_SIZE,
+        FireStoreUtils.MAX_FIRESTORE_BATCH_SIZE,
```
Nice move :)

okotsopoulos commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java:
```diff
@@ -284,7 +286,7 @@ public <T> T runTransactionWithRetry(
         return transactionGet(transactionOp, transaction);
       } catch (Exception ex) {
-        final long retryWait = (long) (SLEEP_BASE_SECONDS * Math.pow(2.5, retry));
+        final long retryWait = Math.min(60, (long) (SLEEP_BASE_SECONDS * Math.pow(2.5, retry)));
```
Lemme guess, the 60 minute cap is coming from hard-won experience? :)

nmalfroy replied:
Yeah, I have a followup PR where I set it to 20.  The problem is that this was growing exponentially so I want to cap it.  I'll make that change here (as well as making it a constant)

okotsopoulos commented on src/test/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDaoTest.java:
```diff
@@ -172,6 +179,161 @@ public void directoryOperationsTest() throws Exception {
+  // Tests that bulk directory ingest works  initially and when there are collisions with the same
+  // load tag and failures when there are collisions with different load tags
+  public void bulkDirectoryOperationsTest() throws Exception {
+    List<String> directories = new ArrayList<>();
```
Rather than clearing out and reusing `directories`, what do you think about defining new lists?  I'd personally find that a bit easier to follow, and then I think you'd be able to initialize them more concisely with `List.of()` (assuming that the mutability was only required to be able to clear the list and add new items).

nmalfroy replied:
yeah, that makes sense.  I'll update

okotsopoulos commented on src/test/java/bio/terra/service/filedata/google/firestore/FireStoreFileDaoTest.java:
```diff
@@ -98,27 +98,36 @@ public void createDeleteFileTest() throws Exception {
-      fileDao.createFileMetadata(firestore, datasetId, fsf);
-      fileList.add(fsf);
-    }
+    List<FireStoreFile> fileList = IntStream.range(0, 5).boxed().map(i -> makeFile()).toList();
```
I think you could use a method reference for the map input if you'd like:

```
…map(this::makeFile).toList();
```

nmalfroy replied:
ah sadly you can't because `makeFile()` doesn't take in a number

okotsopoulos replied:
Oh right!

okotsopoulos commented on src/test/java/bio/terra/service/filedata/google/gcs/GcsPdaoTest.java:
```diff
@@ -158,6 +166,55 @@ public void testGetBlobContentsMatchingPath() {
+          "times between copies is different since file was deleted",
+          gcsPdao.copyFile(dataset, fileLoadModel, fileId, targetBucket).getCreatedDate(),
+          not(equalTo(initialTime)));
+    } catch (InterruptedException e) {
```
I see that this is an existing pattern in these tests, but I wonder if this cleanup instead be handled in an @After method by registering the blob(s) to delete?

nmalfroy replied:
Gah, I think there's some subtle differences with when the different methods are deleting.  I love the idea but if it's ok with you I'm going to leave as is for now

okotsopoulos replied:
Yeah no problem!

okotsopoulos commented on src/test/java/bio/terra/service/filedata/FileIdServiceTest.java:
```diff
@@ -38,6 +38,18 @@ public void testPredictableUUIDEquals() {
   }
 
+  @Test
+  public void testPredictableUUIDWithNotMD5Equals() {
```
It's not clear to me from reading the test alone that these `FSItem`s have no md5 checksums (outside of trusting the test name :)

Can we assert this, if only to signal to the reader that we're testing the condition we expect?

nmalfroy replied:
Good call.  I'll explicitly set the md5 to null

okotsopoulos replied:
Thank you!

Actually, now I'm curious… what's the expected behavior if two files are equal but one has an MD5 and the other doesn't?  Would they get matching IDs?

nmalfroy replied:
right now they wouldn't.  that's probably bad?  I'm going to do a followup PR (with user specified MD5s) that will make this required (and throw a meaningful exception if there is no md5)

snf2ye commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDao.java:
```diff
@@ -81,13 +81,59 @@ public void createDirectoryEntry(Dataset dataset, FireStoreDirectoryEntry newEnt
+   * @throws InterruptedException If something goes wrong talking to Firestore
+   * @throws FileSystemExecutionException If the file already exists but with a different load tag
+   */
+  public void upsertDirectoryEntries(Dataset dataset, String loadTag, List<String> directories)
+      throws InterruptedException {
+    Firestore firestore =
+        FireStoreProject.get(dataset.getProjectResource().getGoogleProjectId()).getFirestore();
+    String datasetId = dataset.getId().toString();
+    directoryDao.upsertDirectoryEntries(firestore, datasetId, loadTag, directories);
+  }
+
+  /**
+   * Write directory entries. If attempting to write a document that has a different ID but the same
+   * load tag, leave the entry untouched and add an entry to the return map. If the load tags don't
+   * match, then throw an exception
+   *
+   * @param dataset The dataset being ingested into
+   * @param newEntries All entries to upsert
+   * @return A map of IDs for entries that already exist. The key is the passed in id and the value
+   *     is the existing value.
+   * @throws InterruptedException If something goes wrong talking to Firestore
+   * @throws FileSystemExecutionException If the file already exists but with a different load tag
+   */
+  public Map<UUID, UUID> upsertDirectoryEntries(
+      Dataset dataset, List<FireStoreDirectoryEntry> newEntries) throws InterruptedException {
+    Firestore firestore =
+        FireStoreProject.get(dataset.getProjectResource().getGoogleProjectId()).getFirestore();
+    String datasetId = dataset.getId().toString();
+    return directoryDao.upsertDirectoryEntries(firestore, datasetId, newEntries);
+  }
```
There appears to be a lot of overlap between these two methods and the directoryDao.upsertDirectoryEntries() methods. The code comments and the naming of these methods do not make it clear when you would use one method over the other. 
The two main differences I'm seeing:
1. Creating the FireStoreDirectoryEntry list in the directoryDao vs. passed in as an argument.  - I think we could the list of string directory names into a list of FireStoreDirectoryEntries in the FirestoreDao so that we could share more of the code in the directoryDao. 
2. Returning list of boolean vs. map of IdConflict records. -- This seems to be the core difference. It would be nice to switch to using one of these return methods, but that's probably out of scope. Could we still share more the logic besides this piece?

nmalfroy replied:
Yeah, I think this is the result of feature creep over the years.  The problem is that we have a way to ingest where we pass the load tag to the endpoint and an ingest where each file has their own load tag (so you could theoretically have multiple load tags per ingest).  That being said I can try to refactor to share more of the code

snf2ye replied:
Awesome, thanks!

okotsopoulos commented on src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkBulkModeResponseStep.java:
```diff
@@ -0,0 +1,31 @@
+// It expects the following working map data:
+// - BULK_LOAD_RESULT - a BulkLoadArrayResultModel object
+//
+public class IngestBulkBulkModeResponseStep implements Step {
```
If you wanted, you could have this extend `DefaultUndoStep` instead, and then not need to override `undoStep` below.

nmalfroy replied:
ooh good call

okotsopoulos commented on src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpBulkFileStep.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.concurrent.ExecutorService;
+import java.util.stream.Stream;
+
+public class IngestBulkGcpBulkFileStep extends IngestBulkGcpStep {
```
`IngestBulkGcpBulkFileStep` is an extra bulky name but I can't think of a leaner one myself :)

nmalfroy replied:
Haha right!?  I've been working with Spring for too long

okotsopoulos commented on src/test/java/bio/terra/common/fixtures/FlightContextFixtures.java:
```diff
@@ -0,0 +1,13 @@
+
+public class FlightContextFixtures {
+
+  public static FlightMap makeContextMap(Map<String, Object> context) {
```
Seeing this method standalone I'm not sure when you'd want to use it, would you mind adding a docstring?

nmalfroy replied:
good call

okotsopoulos commented on src/test/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpStepTest.java:
```diff
@@ -0,0 +1,398 @@
+
+  private static final String LOAD_TAG = "loadtag";
+  private static final UUID PROFILE_ID = UUID.randomUUID();
+  private static final AuthenticatedUserRequest TEST_USER =
```
If you wanted, you could use `AuthenticationFixtures.randomUserRequest()` here!

nmalfroy replied:
oh neat!

okotsopoulos commented on src/main/java/bio/terra/service/filedata/flight/ingest/FileIngestBulkFlight.java:
```diff
@@ -104,17 +108,20 @@ public FileIngestBulkFlight(FlightMap inputParameters, Object applicationContext
     // Parameters dependent on which request we get
     int maxFailedFileLoads;
     UUID profileId;
+    boolean isBulkMode;
```
G'ah, the more I see this the more I find it confusing to have a bulk mode for an ingest that can already have a "bulk" mode.  Naming things is hard!

nmalfroy replied:
haha yeah...

okotsopoulos commented on src/main/resources/api/data-repository-openapi.yaml:
```diff
@@ -4569,6 +4571,17 @@ components:
           description: Array files to load
           items:
             $ref: '#/components/schemas/BulkLoadFileModel'
+        bulkMode:
```
Actually, seeing that we're repeating this definition with its necessary caveats and warnings, it could be nice to define it and reference instead?

nmalfroy replied:
fair point!

okotsopoulos replied:
Thanks for taking a crack at this, bummer that we couldn't keep it given OpenAPI's deficiencies around default values for references!

okotsopoulos commented on src/test/java/bio/terra/integration/FileTest.java:
```diff
@@ -161,10 +176,129 @@ public void longFileLoadTest() throws Exception {
+  // The purpose of these tests is to ingest files using the bulk mode in various permutations
+  @Test
+  public void bulkFileLoadTestTdrHostedRandomIdFile() throws Exception {
+    bulkFileLoadTest(100, false, false, false);
```
What do you think about storing 100 in a static variable `BULK_FILES_TO_LOAD`?  That way it's easier to dial up / down.

nmalfroy replied:
I guess I was thinking this was more for one off local tests and you'd only do one at a time but I think your suggestion makes sense

okotsopoulos commented on src/test/java/bio/terra/integration/FileTest.java:
```diff
@@ -161,10 +176,129 @@ public void longFileLoadTest() throws Exception {
+      loadSummary = dataRepoFixtures.bulkLoad(steward(), datasetId, bulkLoad);
+    }
+
+    logger.info("Ingest took %s milliseconds".formatted(Instant.now().toEpochMilli() - start));
```
Future me thanks you for these logs, this will make debugging and assessing performance easier!

okotsopoulos commented on src/test/java/bio/terra/integration/FileTest.java:
```diff
@@ -161,10 +176,129 @@ public void longFileLoadTest() throws Exception {
+          "bulkFileLoadTest loading " + filesToLoad + " files into dataset id " + datasetId);
+
+      for (int i = 0; i < filesToLoad; i++) {
+        String tailPath = "/fileloadprofiletest/1KBfile.txt";
```
These are paths pointing to real files in GCS, right?  May be personal preference but I'd recommend bumping them up to top-level variables in that case.

nmalfroy replied:
fair enough.  I'll pull them out of the loop.  I do want to keep them close to where they're used though

okotsopoulos reviewed (approved):
I really appreciate the efforts you took to make this PR understandable despite its size.  The combo of our walkthrough plus stepping through the commits made a difference!  The speed improvement is fantastic and I'm looking forward to seeing this in action.

Generally I do worry about the way our ingest processes increase in complexity each time we add a variant.  But I don't think there's anything that could be done immediately to alleviate this, and I do appreciate that your tests are primarily unit / connected rather than exploding out our integration tests further.

I have left a number of suggestions, but none are show-stoppers!

snf2ye commented on src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpStep.java:
```diff
@@ -0,0 +1,512 @@
+      throws InterruptedException {
+    logger.info("Extracting Directories");
+    // Get all directories to create:
+    List<String> directories =
```
Continuing from my previous comment - could we build the list of FireStoreDirectoryEntries here and pass that to fileDao.upsertDirectoryEntries()?

nmalfroy replied:
totally!  I ended up doing that in the `FireStoreDao` class

snf2ye commented on src/main/java/bio/terra/service/filedata/flight/ingest/IngestBulkGcpStep.java:
```diff
@@ -0,0 +1,512 @@
+      // Google's error message suggests retrying the operation
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, ex);
+    } catch (FileSystemAbortTransactionException rex) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, rex);
```
In attempting to compare this step to the existing code for ingest, I see there is a lot of logic in the[ old implementation](https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/filedata/flight/ingest/IngestDriverStep.java#L126) that handles the case on retry where an ingest is in progress or has been orphaned. 

You mentioned today that this flight should be idempotent. How would it handle the case that a file is in progress of being ingested when we retry the ingest? Could we end up with duplicate files?

nmalfroy replied:
That's a great question.  We shouldn't because even if we do regenerate file ids - as long as the paths match, we end up grabbing the id from the path from a previous ingest (that's one of the differences with the bulk upload method you commented about above). That's where the idconflicts map comes into play

nmalfroy committed:
PR Feedback: consolidate upsert directory methods

nmalfroy committed:
PR Feedback: add unit tests / check for directory extractor

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/FileMetadataUtils.java:
```diff
@@ -109,4 +110,25 @@ public static Set<String> findNewDirectoryPaths(
+   * @return An ordered list of paths that could potentially be created. Note: this excludes the
+   *     passed in path
+   */
+  public static List<String> extractDirectoryPaths(String path) {
+    List<String> allPaths = new ArrayList<>();
+    String interimPath = "";
+    allPaths.add("/");
+    for (String part : getDirectoryPath(path).split("/")) {
+      if (!StringUtils.isEmpty(part)) {
+        interimPath += "/" + part;
+        allPaths.add(interimPath);
+      }
+    }
+    return allPaths;
+  }
```
This seems a little wasteful as it uses `split` but then re-allocs each prefix string separately. Using `indexOf()` instead of `split` would avoid that duplication.
```
  public static List<String> extractDirectoryPaths2(String path) {
    List<String> allPaths = new ArrayList<>();
    allPaths.add("/");
    int pos = 1;
    while ((pos = path.indexOf("/", pos)) != -1) {
      allPaths.add(path.substring(0, pos++));
    }
    return allPaths;
  }
```

nmalfroy replied:
Yeah, I just found this way a touch easier to read

nmalfroy committed:
PR Feedback

nmalfroy committed:
Sadly, I can't combine the properties

...because the code generator doesn't handle the defaults for the client

nmalfroy committed:
Truncate the job response if the number of files is too big (1000)
